### PR TITLE
MR1L0, MR2L0 instances of lcls-twincat-optics

### PR DESCRIPTION
**Pull request by jsheppard95**
This pull request builds on that for `pcdshub/lcls-twincat-optics` (https://github.com/pcdshub/lcls-twincat-optics/pull/1) and includes the implementation of this optics PLC library for specific systems, namely MR1L0 and MR2L0.

This project will instantiate `lcls-twincat-optics.FB_RunHOMS` with parameters specific to each mirror system (i.e gantry tolerances, encoder offsets, etc). Development can then follow in parallel for both this project and `lcls-twincat-optics` as needed.

TODO (initial thoughts):
1. Build IO tree for each optics system
2. Set variable links via TcLink
3. Set up pytmc pragmas (both here and in `lcls-twincat-optics`)
4. Sprinkle some MPS FBs on top (both here and in `lcls-twincat-optics`)

### CODE REVIEW

**Comment by klauer** [comment]
Summary gist: https://gist.github.com/klauer/40d9ac74aeb0bcb4a501c4a7d732b244

**Comment by klauer** [comment]
An idea to take with a grain of salt - what about a struct to hold the gantry status/information? These variables are duplicated for each mirror system (M1L0/M2L0):
Take:
```pascal
fbRunHOMS_M1L0 : FB_RunHOMS;
bExecuteCoupleYM1L0 : BOOL;
bExecuteDecoupleYM1L0 : BOOL;
bExecuteCoupleXM1L0 : BOOL;
bExecuteDecoupleXM1L0 : BOOL;
bGantryAlreadyCoupledYM1L0 : BOOL;
bGantryAlreadyCoupledXM1L0 : BOOL;
nCurrGantryYM1L0 : LINT; // encoder counts = nm
nCurrGantryXM1L0 : LINT; // encoder counts = nm
```

Convert to:
```pascal
STRUCT
    fbRunHOMS : FB_RunHOMS;
    bExecuteCoupleY : BOOL;
    bExecuteDecoupleY : BOOL;
    bExecuteCoupleX : BOOL;
    bExecuteDecoupleX : BOOL;
    bGantryAlreadyCoupledY : BOOL;
    bGantryAlreadyCoupledX : BOOL;
    nCurrGantryY : LINT; // encoder counts = nm
    nCurrGantryX : LINT; // encoder counts = nm
END_TYPE
```
I'd guess the pytmc pragmas would be even easier if done this way, though perhaps the TcLink pragmas couldn't be done similarly?

Otherwise, your plan + the current PR as-is looks fine to me.

**Comment by jsheppard95** [comment]
> An idea to take with a grain of salt - what about a struct to hold the gantry status/information? These variables are duplicated for each mirror system (M1L0/M2L0):
> Take:
> 
> ```pascal
> fbRunHOMS_M1L0 : FB_RunHOMS;
> bExecuteCoupleYM1L0 : BOOL;
> bExecuteDecoupleYM1L0 : BOOL;
> bExecuteCoupleXM1L0 : BOOL;
> bExecuteDecoupleXM1L0 : BOOL;
> bGantryAlreadyCoupledYM1L0 : BOOL;
> bGantryAlreadyCoupledXM1L0 : BOOL;
> nCurrGantryYM1L0 : LINT; // encoder counts = nm
> nCurrGantryXM1L0 : LINT; // encoder counts = nm
> ```
> 
> Convert to:
> 
> ```pascal
> STRUCT
>     fbRunHOMS : FB_RunHOMS;
>     bExecuteCoupleY : BOOL;
>     bExecuteDecoupleY : BOOL;
>     bExecuteCoupleX : BOOL;
>     bExecuteDecoupleX : BOOL;
>     bGantryAlreadyCoupledY : BOOL;
>     bGantryAlreadyCoupledX : BOOL;
>     nCurrGantryY : LINT; // encoder counts = nm
>     nCurrGantryX : LINT; // encoder counts = nm
> END_TYPE
> ```
> 
> I'd guess the pytmc pragmas would be even easier if done this way, though perhaps the TcLink pragmas couldn't be done similarly?
> 
> Otherwise, your plan + the current PR as-is looks fine to me.

I definitely like that. So the struct would be defined in a DUT and then `Main` would just instantiate the struct for each mirror system. I see what you mean in that I would only need one set of pragmas but linking via `TcLink` would be tricky. I'll give it a go though and see what it looks like.

File: MR1L0-MR2L0/MR1L0_MR2L0_PLC/POUs/Main.TcPOU [hunk 1]
```diff
@@ -0,0 +1,147 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4022.12">
+  <POU Name="Main" Id="{371c560d-bfd7-44d3-85a8-59ca1fa1938a}" SpecialFunc="None">
+    <Declaration><![CDATA[PROGRAM Main
+VAR
+	// Axes for each system
+	// M1L0
+	M1 : DUT_MotionStage:=(nCommand:=ENUM_EpicsMotorCmd.EPICS_MOTOR_MOVE_ABS); // Yup
```
Comment (klauer) [inline MR1L0-MR2L0/MR1L0_MR2L0_PLC/POUs/Main.TcPOU]: Reminder: you'll need to switch the pragmas to `DUT_MotionStage` - currently they're on `FB_MotionStage` and as such the PVs are incorrect in the `st.cmd`.

**Comment by klauer** [comment]
Interesting - looks like the TcLinkTo pragma can be done with something like [this](https://infosys.beckhoff.com/english.php?content=../content/1033/tc3_plc_intro/9007202362715915.html&id=):
```
 
    {attribute 'TcLinkTo' := '.bIn  := TIID^Device 1 (EtherCAT)^Term 1 (EK1100)^Term 3 (EL1008)^Channel 2^Input;
                              .bOut := TIID^Device 1 (EtherCAT)^Term 1 (EK1100)^Term 2 (EL2008)^Channel 2^Output'}
    // Link when using function blocks (FB_Modules with allocated input bIn and allocated output bOut)
    fbVar4         : FB_Module;
 
```

File: MR1L0-MR2L0/MR1L0_MR2L0_PLC/POUs/Main.TcPOU (hunk 1 above)
Comment (jsheppard95) [reply MR1L0-MR2L0/MR1L0_MR2L0_PLC/POUs/Main.TcPOU]: Resolved in commit 4aec284

**Comment by klauer** [comment]
When you get a chance @jsheppard95 (no rush at all), I think it's time we merge this and open up a new PR.

**Comment by ZLLentz** [comment]
Is this ready for review?

**Comment by jsheppard95** [comment]
Yes I would love some review. Commit eef9274 is currently deployed, but I would love to get this merged and then open up new PR's (that follow our normal Git workflow) for things like logging, MPS, etc.

**Comment by ZLLentz** [comment]
49,741 lines added, 0 lines removed, 99 files changed

I'm going in

File: MR1L0-MR2L0/MR1L0-MR2L0.tsproj [hunk 2]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
```
Comment (ZLLentz) [inline MR1L0-MR2L0/MR1L0-MR2L0.tsproj]: There is a second file with a `.tsproj.archive` extension that is empty. Is the second file intentional? Or should it be removed?

File: MR1L0-MR2L0/MR1L0_MR2L0_PLC/GVLs/GVL_SerialIO.TcGVL [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  <GVL Name="GVL_SerialIO" Id="{e2338c89-1965-407c-b604-590f80e0dc24}">
+    <Declaration><![CDATA[VAR_GLOBAL
+	//Better have your inputs and outputs!
+	// M1L0
+	Serial_stComIn_M1L0   AT %I*	:	EL6inData22B (*KL6inData22B*);
+	Serial_stComOut_M1L0  AT %Q*	:	EL6outData22B (*KL6outData22B*);
+
+	// M2L0
+	Serial_stComIn_M2L0   AT %I*	:	EL6inData22B (*KL6inData22B*);
+	Serial_stComOut_M2L0  AT %Q*	:	EL6outData22B (*KL6outData22B*);
```
Comment (ZLLentz) [inline MR1L0-MR2L0/MR1L0_MR2L0_PLC/GVLs/GVL_SerialIO.TcGVL]: Are the comments here copied from the Beckhoff serial coms example?

File: MR1L0-MR2L0/MR1L0_MR2L0_PLC/MR1L0_MR2L0_PLC.plcproj [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+  </ItemGroup>
+  <ItemGroup>
+    <LibraryReference Include="LCLS General,0.0.0,SLAC">
+      <Namespace>LCLS_General</Namespace>
+    </LibraryReference>
+    <LibraryReference Include="lcls-twincat-motion,1.1.0,SLAC">
+      <Namespace>lcls_twincat_motion</Namespace>
+    </LibraryReference>
+    <LibraryReference Include="lcls-twincat-optics,0.0.0,SLAC">
+      <Namespace>lcls_twincat_optics</Namespace>
```
Comment (ZLLentz) [inline MR1L0-MR2L0/MR1L0_MR2L0_PLC/MR1L0_MR2L0_PLC.plcproj]: Don't forget to switch to tagged versions when things stabilize. If the time isn't "right now", then you should slap an issue on this repo to remind you to switch it later.

File: MR1L0-MR2L0/MR1L0_MR2L0_PLC/POUs/Main.TcPOU [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+// Convert nCurrGantry to um (smaller number) to read out in epics
+M1L0.fCurrGantryY_um := LINT_TO_REAL(M1L0.nCurrGantryY) / 1000.0;
+M1L0.fCurrGantryX_um := LINT_TO_REAL(M1L0.nCurrGantryX) / 1000.0;
+
+// FB_MotionStage's for non-piezo axes
+fbMotionStage_m1(stMotionStage:=M1);
+fbMotionStage_m2(stMotionStage:=M2);
+fbMotionStage_m3(stMotionStage:=M3);
+fbMotionStage_m4(stMotionStage:=M4);
+fbMotionStage_m6(stMotionStage:=M6);
```
Comment (ZLLentz) [inline MR1L0-MR2L0/MR1L0_MR2L0_PLC/POUs/Main.TcPOU]: Make sure you didn't double up on these in the optics lib. They are only intended to be called once per axis.

File: MR1L0-MR2L0/MR1L0_MR2L0_PLC/POUs/Serial Comm/P_Serial_Com.TcPOU [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+fbSerialLineControl_EL6001_M1L0(Mode:= SERIALLINEMODE_EL6_22B (*SERIALLINEMODE_KL6_22B_STANDARD*),
+						        pComIn:= ADR(Serial_stComIn_M1L0),
+						   	    pComOut:=ADR(Serial_stComOut_M1L0),
+						        SizeComIn:= SIZEOF(Serial_stComIn_M1L0),
+						        TxBuffer:= Serial_TXBuffer_M1L0,
+						        RxBuffer:= Serial_RXBuffer_M1L0,
+						        Error=> ,
+						        ErrorID=> );
+
+fbSerialLineControl_EL6001_M2L0(Mode:= SERIALLINEMODE_EL6_22B (*SERIALLINEMODE_KL6_22B_STANDARD*),
```
Comment (ZLLentz) [inline MR1L0-MR2L0/MR1L0_MR2L0_PLC/POUs/Serial Comm/P_Serial_Com.TcPOU]: More hanging comments from the serial coms example

File: iocBoot/ioc-MR1L0-MR2L0-PLC/Makefile [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+# Set PRODUCTION_IOC to 1 to move from a testing to a production IOC:
+PRODUCTION_IOC := 1
+
+PROJECT_NAME := MR1L0-MR2L0
+PROJECT_PATH := ../../MR1L0-MR2L0/MR1L0-MR2L0.tsproj
+PLC := MR1L0_MR2L0_PLC
+
+PYTMC_OPTS := 
+PREFIX := PREFIX
```
Comment (ZLLentz) [inline iocBoot/ioc-MR1L0-MR2L0-PLC/Makefile]: You should change your prefix to something meaningful

**Comment by jsheppard95** [comment]
@ZLLentz You are a hero, I know this was a monster of a PR. I'll go through/address each comment (much appreciated) and then make a new commit.

**Comment by ZLLentz** [comment]
@jsheppard95 I decided against going 100% and opening this PR on my twincat though, on the grounds that you say it is working. I was focusing on "small things"

**Comment by jsheppard95** [comment]
That's fair. And in general I have found no major bugs in this project. I think the most pressing issue is that described here: https://github.com/pcdshub/lcls-twincat-optics/issues/5

But this is more specific to `lcls-twincat-optics`

File: MR1L0-MR2L0/MR1L0_MR2L0_PLC/MR1L0_MR2L0_PLC.plcproj (hunk 4 above)
Comment (jsheppard95) [reply MR1L0-MR2L0/MR1L0_MR2L0_PLC/MR1L0_MR2L0_PLC.plcproj]: Yes I just want to get the piezo-control branch merged before releasing the first version. Just made Issue https://github.com/pcdshub/MR1L0-MR2L0/issues/2

File: MR1L0-MR2L0/MR1L0_MR2L0_PLC/POUs/Main.TcPOU (hunk 5 above)
Comment (jsheppard95) [reply MR1L0-MR2L0/MR1L0_MR2L0_PLC/POUs/Main.TcPOU]: Just checked to confirm and they are only called here, except for that of the pitch axis which is called in `lcls-twincat-optics.FB_PitchControl`.

File: MR1L0-MR2L0/MR1L0-MR2L0.tsproj (hunk 2 above)
Comment (jsheppard95) [reply MR1L0-MR2L0/MR1L0-MR2L0.tsproj]: Not intentional and unsure how it got there, went ahead and deleted it in commit 
ced9328

File: MR1L0-MR2L0/MR1L0_MR2L0_PLC/GVLs/GVL_SerialIO.TcGVL (hunk 3 above)
Comment (jsheppard95) [reply MR1L0-MR2L0/MR1L0_MR2L0_PLC/GVLs/GVL_SerialIO.TcGVL]: Yes they were, changed them to something more useful in commit ced9328

File: MR1L0-MR2L0/MR1L0_MR2L0_PLC/POUs/Serial Comm/P_Serial_Com.TcPOU (hunk 6 above)
Comment (jsheppard95) [reply MR1L0-MR2L0/MR1L0_MR2L0_PLC/POUs/Serial Comm/P_Serial_Com.TcPOU]: Resolved in commit ced9328

File: iocBoot/ioc-MR1L0-MR2L0-PLC/Makefile (hunk 7 above)
Comment (jsheppard95) [reply iocBoot/ioc-MR1L0-MR2L0-PLC/Makefile]: Thanks for catching that, just changed to `PLC:LFE:OPTICS` in commit ced9328

**Comment by jsheppard95** [comment]
Thanks for the comments, if there no other obvious issues I think this is ready to be merged. I'd just like to change the project name to `lcls-plc-lfe-optics` and the PLC name to `lfe_optics`, and then follow this standard through the other directories. This seems to follow the convention of other PLC projects and it more clearly corresponds to the PLC's network info: netconfig = `plc-lfe-optics`, hostname=`LFE-OPTICS`, etc.

I just fear the name changes will be a TwinCAT challenge...

**Comment by jsheppard95** [comment]
OK changing the project name really wasn't that bad actually. I think I have the desired file structure now, though I did need to remove the scope (unclear if that should have been version controlled in the first place).

@ZLLentz How would you recommend going about merging? As one would expect, each file now shows up as deleted and there are new untracked files under the project name, that's not so bad. Should I just rename `pcdshub/MR1L0-MR2L0` to `pcdshub/lcls-plc-lfe-optics` along with my fork, and then just push the project with new names to my fork? And then of course I'll need to rebuild the ioc.

**Comment by ZLLentz** [comment]
do the github renaming last, just focus on all the changes in your project:

git add <new directory>
git remove <old directory>
git commit

**Comment by ZLLentz** [comment]
For the github renaming, yes just change the names in this repo's settings and in your fork's settings and update your remote urls. No reason to do that before merging this PR.

**Comment by jsheppard95** [comment]
I decided to leave logging out for now. It looks like the library version on the prog node is from @slacAWallace 's fork: https://github.com/slacAWallace/lcls-twincat-general/tree/ethercat-logging
And since I'll have to change the version after it gets merged anyway, figured I might as well wait until after the `lcls-twincat-general` merge.

File: iocBoot/ioc-lfe-optics/envPaths [hunk 8]
```diff
@@ -9,5 +9,5 @@ epicsEnvSet("MOTOR","/reg/g/pcds/epics//R7.0.2-2.0/modules/motor/R6.9-ess-0.0.1"
 epicsEnvSet("ASYN","/reg/g/pcds/epics//R7.0.2-2.0/modules/asyn/R4.35-0.0.1")
 epicsEnvSet("ADS","/reg/g/pcds/epics//R7.0.2-2.0/modules/twincat-ads/R2.0.0-0.0.5")
 epicsEnvSet("EPICS_BASE","/reg/g/pcds/epics//base/R7.0.2-2.0")
-epicsEnvSet("IOC_TOP", "/reg/neh/home5/sheppard/HOMS/ioc/MR1L0-MR2L0/iocBoot/ioc-MR1L0-MR2L0-PLC")
-epicsEnvSet("IOC_DATA", "/reg/d/iocData")
+epicsEnvSet("IOC_TOP", "/reg/g/pcds/epics/ioc/lcls-plc-lfe-optics/iocBoot/ioc-lfe-optics")
+epicsEnvSet("IOC_DATA", "/root/iocData")
```
Comment (ZLLentz) [inline iocBoot/ioc-lfe-optics/envPaths]: take care that this gets set correctly when you build the IOC

File: iocBoot/ioc-lfe-optics/st.cmd [hunk 9]
```diff
@@ -1,11 +1,11 @@
...
+#!/reg/g/pcds/epics/ioc/common/ads-ioc/R0.2.0/bin/rhel7-x86_64/adsIoc
 
 < envPaths
 
 epicsEnvSet("ADS_IOC_TOP", "$(TOP)" )
 
-epicsEnvSet("IOCNAME", "ioc-MR1L0-MR2L0-PLC" )
-epicsEnvSet("ENGINEER", "sheppard" )
+epicsEnvSet("IOCNAME", "ioc-lfe-optics" )
+epicsEnvSet("ENGINEER", "root" )
```
Comment (ZLLentz) [inline iocBoot/ioc-lfe-optics/st.cmd]: I am root

File: iocBoot/ioc-lfe-optics/envPaths (hunk 8 above)
Comment (jsheppard95) [reply iocBoot/ioc-lfe-optics/envPaths]: Right so I would've thought that I need to create a directory `/reg/d/iocData/ioc-lfe-optics` with sub-directories `archive`, `autosave`, `iocInfo`, and `logs` (as this is what I had to do using the old name `ioc-MR1L0-MR2L0-PLC`, but somehow these already exist with empty sub-directories except `iocInfo` which is fille with log files. These directories could have been made by the `iocmanager` as I did call my IOC `ioc-lfe-optics`.

Comment (jsheppard95) [reply iocBoot/ioc-lfe-optics/envPaths]: Just changed `epicsEnvSet("IOC_DATA", "/root/iocData")` in commit 2ac760a, but does this actually happen automatically when I build on psb7?

File: iocBoot/ioc-lfe-optics/st.cmd (hunk 9 above)
Comment (jsheppard95) [reply iocBoot/ioc-lfe-optics/st.cmd]: Oh good catch, fixed that in commit 2ac760a along with the extra `PREFIX` fields, but again I'm guessing this happens automtically when I build on psb7 after setting the `PREFIX` field in my `Makefile`?

File: iocBoot/ioc-lfe-optics/envPaths (hunk 8 above)
Comment (ZLLentz) [reply iocBoot/ioc-lfe-optics/envPaths]: This should happen automatically when you build on psbuild

File: iocBoot/ioc-lfe-optics/st.cmd (hunk 9 above)
Comment (ZLLentz) [reply iocBoot/ioc-lfe-optics/st.cmd]: Yes

**Comment by ZLLentz** [comment]
I think this looks good! Are you ready to put big PR to rest?

**Comment by jsheppard95** [comment]
Yes lets do it! Should I go ahead and merge it in?

**Pull request merged**